### PR TITLE
Profiling CircleCI job

### PR DESCRIPTION
**Pull request by davidor**
This PR introduces a simple script to profile APIcast and detect whether it's using LuaJIT traces in the majority of requests or not.

The script basically performs the checks I did manually in https://github.com/3scale/apicast/pull/937. Running this script in CircleCI will prevent us from accidentally adding code that's not jittable.

For now, there's just one basic test. It starts an APIcast in echo-mode and a second APIcast that just has an Upstream policy that matches any path and proxies the request to the APicast in echo mode. Then, it profiles APIcast using the `ngx-lj-trace-exits` tool.

The idea is to add more tests in the future. Those tests will detect other kinds of problems with other profiling tools, but also cover more complex scenarios using several policies, etc.

For now, the code makes a lot of assumptions, but I think that's fine. We can make it more generic as we add more functionality.

The only downside that I see is that the new "profiling" job takes a lot of time in CircleCI, around 10 minutes.

### CODE REVIEW

File: script/profiling/profile.lua [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    start_echo_apicast()
+    start_apicast_to_profile()
+
+    local master_pid = pid_from_file(profile_apicast.pid_file)
+    return capture_output('pgrep -P ' .. master_pid)
+  end
+
+  function SimpleScenario.teardown()
+    local pids = {
+      pid_from_file(profile_apicast.pid_file),
```
Comment (mikz) [inline script/profiling/profile.lua]: You can use `jobs -l` to list all the background jobs and terminate the jobs. You can also use`kill %1` to kill them by the relative ID, not pid.

Comment (mikz) [reply script/profiling/profile.lua]: Also possibly `pkill -P $$` should kill all the child processes.

File: script/profiling/profile.lua [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+local function pid_from_file(pid_file)
+  return capture_output("cat " .. pid_file)
+end
+
+local Profiler = {}
+do
+  function Profiler.check_trace_exits(pid, duration)
+    local trace_exits_command = string.format(
+      'sudo PATH="$PATH:/usr/local/bin" /usr/local/bin/ngx-lj-trace-exits.sxx -x %s --arg time=%s',
```
[the remark this answers is not in the document]
Comment (davidor) [reply script/profiling/profile.lua]: Good catch. Fixed :+1:

File: script/profiling/profile.lua [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      return false, 'only ' .. perc_using_traces .. '% of the requests used LuaJIT traces.'
+    else
+      return true
+    end
+  end
+end
+
+local Wrk = {
+  make_requests = function(duration)
+    local command = string.format('wrk -c 10 -t 10 -d %s http://localhost:8080 &', duration)
```
[the remark this answers is not in the document]
Comment (davidor) [reply script/profiling/profile.lua]: Agreed. Fixed :+1:

File: script/profiling/profile.lua (hunk 1 above)
Comment (davidor) [reply script/profiling/profile.lua]: I think that using the PID gives us more control. With `jobs -l`, there could be others already running or we might introduce other background jobs in the future and the indexes could change.

File: script/profiling/profile.lua [hunk 4]
```diff
@@ -0,0 +1,137 @@
+-- This program only works in an environment that has installed the Openresty
+-- profiling tools using the scripts under /script/install.
+
+local function capture_output(command)
+  local f = assert(io.popen(command))
+
+  local out = {}
+
+  for line in f:lines() do
```
[the remark this answers is not in the document]
Comment (davidor) [reply script/profiling/profile.lua]: It's using plain lua for the script, which is fine, but maybe I should switch to resty for consistency.

[the remark this answers is not in the document]
Comment (davidor) [reply script/profiling/profile.lua]: By the way, I tried running the script with resty and everything seems OK. I also tried running that method in an APIcast policy and it works too. What's the issue with using `io.lines()` ?

[the remark this answers is not in the document]
Comment (mikz) [reply script/profiling/profile.lua]: Normal Lua (5.1) is fine I guess, but it means you can't use any of the 5.2 features or dependencies.

https://github.com/openresty/lua-nginx-module/issues/1346
https://github.com/openresty/resty-cli/issues/35

Not sure if some of the fixes made it to the current release.

[the remark this answers is not in the document]
Comment (davidor) [reply script/profiling/profile.lua]: I'd say Lua 5.1 is fine for now. We can always change that in the future when needed.

**Pull request merged**